src/shared/install.c

```diff
@@ -96,110 +97,27 @@ static int unit_file_is_generated(const LookupPaths *p, const char *path) {
-                else
-                        p = path_join(root_dir, SYSTEM_CONFIG_UNIT_PATH, NULL);
-                break;
-
-        case UNIT_FILE_GLOBAL:
-
-                if (root_dir)
-                        return -EINVAL;
-
-                if (runtime)
-                        p = strdup("/run/systemd/user");
-                else
-                        p = strdup(USER_CONFIG_UNIT_PATH);
-                break;
-
-        case UNIT_FILE_USER:
-
-                if (root_dir)
-                        return -EINVAL;
-
-                if (runtime)
-                        r = user_runtime_dir(&p);
-                else
-                        r = user_config_home(&p);
-                if (r < 0)
-                        return r;
-                if (r == 0)
-                        return -ENOENT;
-
-                break;
-
-        default:
-                assert_not_reached("Bad scope");
-        }
-
-        if (!p)
-                return -ENOMEM;
-
-        *ret = p;
-        return 0;
-}
-
-static bool is_config_path(UnitFileScope scope, const char *path) {
-        int r;
+static int path_is_config(const LookupPaths *p, const char *path) {
+        _cleanup_free_ char *parent = NULL;
 
-        assert(scope >= 0);
-        assert(scope < _UNIT_FILE_SCOPE_MAX);
+        assert(p);
         assert(path);
 
-        /* Checks whether the specified path is intended for
-         * configuration or is outside of it */
-
-        switch (scope) {
-
-        case UNIT_FILE_SYSTEM:
-        case UNIT_FILE_GLOBAL:
-                return path_startswith(path, "/etc") ||
-                        path_startswith(path, SYSTEM_CONFIG_UNIT_PATH) ||
-                        path_startswith(path, "/run");
-
-
-        case UNIT_FILE_USER: {
-                _cleanup_free_ char *p = NULL;
-
-                r = user_config_home(&p);
-                if (r < 0)
-                        return r;
-                if (r > 0 && path_startswith(path, p))
-                        return true;
-
-                p = mfree(p);
+        /* Checks whether the specified path is intended for configuration or is outside of it. We check both the
+         * top-level directory and the one actually configured. The latter is particularly relevant for cases where we
+         * operate on a root directory. */
 
-                r = user_runtime_dir(&p);
-                if (r < 0)
-                        return r;
-                if (r > 0 && path_startswith(path, p))
-                        return true;
+        if (path_startswith(path, "/etc") || path_startswith(path, "/run"))
+                return true;
 
-                return false;
-        }
+        parent = dirname_malloc(path);
+        if (!parent)
+                return -ENOMEM;
 
-        default:
-                assert_not_reached("Bad scope");
-        }
+        return path_equal(parent, p->persistent_config) ||
+                path_equal(parent, p->runtime_config);
```

Comment:
i thought about this for a while, and came to the conclusion that we shouldn't do this here, since this is about search paths really, and using startswith() would mean it's not anymore... but i am a bit torn about this i must say...